central/globaldb/v2backuprestore/restore/postgres.go

```diff
@@ -0,0 +1,167 @@
+	// Set the options for pg_dump from the connection config
+	options := []string{
+		"-d",
+		restoreDB,
+		"--no-owner",
+		"--clean",
+		"--if-exists",
+		"--exit-on-error",
+		"-Fc",
+		"-vvv",
+		"--single-transaction",
+	}
+
+	// Get the common DB connection info
+	options = append(options, pgadmin.GetConnectionOptions(config)...)
+
+	cmd := exec.Command("pg_restore", options...)
+
+	// Set stdin to be the incoming reader
+	cmd.Stdin = fileReader
+
+	pgadmin.SetPostgresCmdEnv(cmd, sourceMap, config)
+	err := pgadmin.ExecutePostgresCmd(cmd)
+
+	if err != nil {
+		// Clean up the restore DB since the restore failed
+		_ = pgadmin.DropDB(sourceMap, config, restoreDB)
+		return errors.Wrap(err, "Unable to restore the postgres dump")
+	}
+
+	return nil
+}
+
+// SwitchToRestoredDB - switches the restore DB to be the active DB
+func SwitchToRestoredDB(sourceMap map[string]string, config *pgxpool.Config) error {
+	log.Info("Switching to restored database")
+
+	// Connect to different database for admin functions
+	connectPool := adminPool(config)
+
+	// Restore succeeded to the separate DB, so we need to drop the original in order to rename
+	// the new one.
+	err := pgadmin.DropDB(sourceMap, config, config.ConnConfig.Database)
+	if err != nil {
+		log.Errorf("Could not drop the DB: %v", err)
+		return err
+	}
+
+	// rename central_restore to postgres
+	err = pgadmin.RenameDB(connectPool, restoreDB, config.ConnConfig.Database)
+	if err != nil {
+		log.Errorf("Could not rename the DB: %v", err)
+		return err
+	}
+
+	// Close the admin connection pool
+	connectPool.Close()
+
+	return nil
+}
+
+func adminPool(config *pgxpool.Config) *pgxpool.Pool {
+	var err error
+	var postgresDB *pgxpool.Pool
+
+	// Clone config to connect to template DB
+	tempConfig := config.Copy()
+
+	// Need to connect on a static DB so we can rename the used DBs.
+	tempConfig.ConnConfig.Database = connectDB
+
+	if err := retry.WithRetry(func() error {
+		postgresDB, err = pgxpool.ConnectConfig(context.Background(), tempConfig)
+		return err
+	}, retry.Tries(postgresOpenRetries), retry.BetweenAttempts(func(attempt int) {
+		time.Sleep(postgresTimeBetweenRetries)
+	}), retry.OnFailedAttempts(func(err error) {
+		log.Errorf("open database: %v", err)
+	})); err != nil {
+		log.Fatalf("Timed out trying to open database: %v", err)
+	}
+
+	return postgresDB
+}
+
+// CheckIfRestoreDBExists - checks to see if a restore database exists
+func CheckIfRestoreDBExists(config *pgxpool.Config) bool {
+	log.Info("CheckIfRestoreDBExists")
+
+	// Connect to different database for admin functions
+	connectPool := adminPool(config)
+
+	existsStmt := "SELECT EXISTS(SELECT 1 FROM pg_catalog.pg_database WHERE datname = $1)"
+
+	row := connectPool.QueryRow(context.Background(), existsStmt, restoreDB)
+	var exists bool
+	if err := row.Scan(&exists); err != nil {
+		return false
+	}
+
+	// Close the admin connection pool
+	connectPool.Close()
```

Comment:
We should defer this from above right? We always want to close this even if we return false?

Reply:
ah, yeah I need to get in the habit of using that go feature.